src/blog/pointer-compression.md

```diff
@@ -0,0 +1,542 @@
+
+Let’s take a closer look at what tagged values are.
+
+## Value tagging in V8
+
+JavaScript values in V8 are represented as objects and allocated on the V8 heap, no matter if they are objects, arrays, numbers or strings. This allows us to represent any value as a pointer to an object.
+
+Many JavaScript programs perform calculations on integer values, such as incrementing an index in a loop. To avoid us having to allocate a new number object each time an integer is incremented, V8 uses the well-known [pointer tagging](https://en.wikipedia.org/wiki/Tagged_pointer) technique to store additional or alternative data in V8 heap pointers.
+
+The tag bits serve a dual purpose: they signal either strong/weak pointers to objects located in V8 heap, or a small integer. Hence, the value of an integer can be stored directly in the tagged value, without having to allocate additional storage for it.
+
+V8 always allocates objects in the heap at word-aligned addresses, which allows it to use the 2 (or 3, depending on the machine word size) least significant bits for tagging. On 32-bit architectures, V8 uses the least significant bit to distinguish Smis from heap object pointers. For heap pointers, it uses the second least significant bit to distinguish strong references from weak ones:
+
+<pre>
+                        |----- 32 bits -----|
+Pointer:                |_____address_____<b>w1</b>|
+Smi:                    |___int31_value____<b>0</b>|
+</pre>
+
+where *w* is a bit used for distinguishing strong pointers from the weak ones.
+
+Note that a Smi value can only carry a 31-bit payload, including the sign bit. In the case of pointers, we have 30 bits that can be used as a heap object address payload. Due to word alignment, the allocation granularity is 4 bytes, which gives us 4 GB of addressable space.
+
+On 64-bit architectures V8 values look like this:
+
+<pre>
+            |----- 32 bits -----|----- 32 bits -----|
+Pointer:    |________________address______________<b>w1</b>|
+Smi:        |____int32_value____|000000000000000000<b>0</b>|
+</pre>
+
+You may notice that unlike 32-bit architectures, on 64-bit architectures V8 can use 32 bits for the Smi value payload. The benefits of 32-bit Smis will be discussed in the following sections.
+
+You may notice that unlike 32-bit architectures, on 64-bit architectures V8 can use 32 bits for the Smi value payload. The implications of 32-bit Smis on pointer compression will be discussed in the following sections.
+
+## Compressed tagged values and new heap layout
+
+With Pointer Compression, our goal is to somehow fit both kinds of tagged values into 32 bits on 64-bit architectures. We can fit pointers into 32 bits by:
+    - making sure all V8 objects are allocated within a 4GB memory range
```

Comment:
@santiaboy Looks like this one got missed and still doesn't render like list (btw, the indentation is also a bit excessive).

Reply:
Done.
The indentation was due to npm run lint. I was originally using " *" and I guess that confused it because it sugested using "-" and 4 spaces.